### PR TITLE
refactor: cdp migration - processing pending tasks in bgQ 

### PR DESCRIPTION
**Pull request by ami-aman**
closes: https://github.com/customerio/issues/issues/11649

### What does this PR do?

This pull request introduces a method that analyzes the Journey's SDK background queue to determine if any tasks from the Journeys SDK remain unexecuted. Should such tasks exist, it then fetches each individual task and executes them one by one. This ensures that all pending tasks are executed prior to transitioning to the CDP and all the pending tasks are sent to the CDP. Once a task is successfully executed, it is removed from the background queue.

This PR currently handles :
- fetching pending tasks
- executing them
- sending data to workspace via (migrated) cdp module 
- added TODO note to the methods that are unavailable currently in segment and have to be added in our CDP SDK 
- Some edge cases like sending data of non-current identifier
- Code cleanup 
- Tested with large data (> 300)
- Automated test with large data (3000takss)

Pending:
- NA

Complete each step to get your pull request merged in. [Learn more about the workflow this project uses](https://github.com/customerio/customerio-ios/blob/develop/docs/dev-notes/GIT-WORKFLOW.md). 
- [ ] [Assign members of your team](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to review the pull request. 
- [ ] Wait for pull request status checks to complete. If there are problems, fix them until you see that [all status checks are passing](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fsymfony.com%2Fdoc%2F4.3%2F_images%2Fdocs-pull-request-symfonycloud.png&f=1&nofb=1). 
- [ ] Wait until the pull request has been reviewed *and approved* by a teammate
- [ ] After code reviews are approved and you determine this PR is ready to merge, select *Squash and Merge* button on this screen, leave the title and description to the default values, then merge the PR.

### CODE REVIEW

**Review by levibostian — COMMENTED** [review]
Could you share some more information about this PR? What work has been done and what has not been done? 

Without this context, I am unsure what to review. 

Quickly reviewing it now, tests need to be added, implementation is incomplete, and there are performance issues in this implementation. 

I understand if more PRs will be made to address these concerns but with the information I have right now, I am unsure if that's true.

**Comment by ami-aman** [comment]
> Could you share some more information about this PR? What work has been done and what has not been done?
> 
> Without this context, I am unsure what to review.
> 
> Quickly reviewing it now, tests need to be added, implementation is incomplete, and there are performance issues in this implementation.
> 
> I understand if more PRs will be made to address these concerns but with the information I have right now, I am unsure if that's true.

@levibostian Thanks for your feedback. I have updated the description with some detail. I hope it helps. Also to address the issue you highlighted with [profileAttributes](https://github.com/customerio/customerio-ios/commit/6f8e6c70245b15ea92183c0635e6d8b11179b92b), I have considered that and updated the code. Let me know if you have any questions.

File: Sources/Common/Util/JsonAdapter.swift [hunk 1]
```diff
@@ -86,6 +86,17 @@ public class JsonAdapter {
         return nil
     }
 
+    public func convertToStringAnyDictionary(_ jsonString: String) -> [String: Any]? {
```
[the remark this answers is not in the document]
Comment (ami-aman) [reply Sources/Common/Util/JsonAdapter.swift:89]: I will add test cases in the end to save time and efforts.

File: Sources/Common/Background Queue/Queue.swift [hunk 2]
```diff
@@ -125,6 +139,28 @@ public class CioQueue: Queue {
...
+
+        return (task.data, queueTaskType)
+    }
+
+    public func deleteProcessedTask(_ task: QueueTaskMetadata) {
+        let storageId = task.taskPersistedId
+        if !storage.delete(storageId: storageId) {
+            logger.error("Failed to delete task with storage id: \(storageId).")
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ami-aman) [reply Sources/Common/Background Queue/Queue.swift:162]: That is one option I did think of but this does not cover a scenario where the task isn't processed for some reason but then deleting the entire directory would remove those tasks too. Does this make sense?

File: Sources/Tracking/CustomerIO.swift [hunk 3]
```diff
@@ -42,6 +42,13 @@ public extension CustomerIO {
         let newDiGraph = DIGraph(sdkConfig: sdkConfig)
         let implementation = CustomerIOImplementation(diGraph: newDiGraph)
 
+        // Check if any unprocessed tasks are pending in the background queue.
+        // If so, iterate over them and process each one.
+        if let allStoredTasks = implementation.getAllStoredTasks(), !allStoredTasks.isEmpty {
+            allStoredTasks.forEach { task in
+                implementation.getStoredTask(for: task)
+            }
+        }
```
[the remark this answers is not in the document]
Comment (ami-aman) [reply Sources/Tracking/CustomerIO.swift:51]: I agree with you, Levi ! I do believe that this _could_ create performance issues for bigger queues and it is a good idea to move this processing to a separate background thread. FYI, as mentioned in the description, it is a TODO to test this implementation with large data !

File: Sources/Tracking/Service/Request/TrackRequestBody.swift [hunk 4]
```diff
@@ -12,3 +12,9 @@ enum EventType: String, Codable {
     case event
     case screen
 }
+
+// TODO: Is this the right place?
+public struct TrackEventTypeForAnalytics: Codable {
```
Comment (Shahroz16) [inline Sources/Tracking/Service/Request/TrackRequestBody.swift:17]: what is this event for?

Comment (ami-aman) [reply Sources/Tracking/Service/Request/TrackRequestBody.swift:17]: This struct is used [here](https://github.com/customerio/customerio-ios/blob/dd054e015ba2a0f1462a1d3e9eb705e4fa6dccde/Sources/Tracking/CustomerIOImplementation.swift#L356) to convert the attributed string data to json. Since we do not have a data model to cater this so I created a new one here. Does it help ?

File: Sources/DataPipeline/DataPipelineImplementation.swift [hunk 5]
```diff
@@ -252,3 +251,66 @@ extension DataPipelineImplementation {
         return attributesPlugin
     }
 }
+
+// To process pending tasks in background queue
+extension DataPipelineImplementation {
+    func processIdentifyFromBGQ(identifier: String, body: [String: Any]?) {
```
Comment (Shahroz16) [inline Sources/DataPipeline/DataPipelineImplementation.swift:257]: should we be adding timestamp as well?

File: Apps/APN-UIKit/APN UIKit/View/LoginViewController.swift [hunk 6]
```diff
@@ -92,9 +92,13 @@ class LoginViewController: BaseViewController {
...
             body = ["first_name": name]
         }
-        CustomerIO.shared.identify(identifier: emailId, body: body)
         storage.userEmailId = emailId
         loginRouter?.routeToDashboard()
+        guard let data = body else {
+            CustomerIO.shared.identify(identifier: emailId)
+            return
+        }
+        CustomerIO.shared.identify(identifier: emailId, body: data)
```
Comment (ami-aman) [inline Apps/APN-UIKit/APN UIKit/View/LoginViewController.swift:101]: This fixes a crash that happens when this method sends a `nil` data to `identify` that CDP doesn't expect.

File: Sources/DataPipeline/DataPipelineImplementation.swift (hunk 5 above)
Comment (ami-aman) [reply Sources/DataPipeline/DataPipelineImplementation.swift:257]: `identify` does not have a `timestamp` in JSON that we get from the queue.
`{"attributes_json_string":"null","identifier":"xyz@customer.io"}`

All event and screen tracks have `timestamp` and have been added to the methods (Refer [this](https://github.com/customerio/customerio-ios/blob/dc754410218d72d9d8d3f87283889cbb1409f8d0/Sources/DataPipeline/DataPipelineImplementation.swift#L271) and [this](https://github.com/customerio/customerio-ios/blob/dc754410218d72d9d8d3f87283889cbb1409f8d0/Sources/DataPipeline/DataPipelineImplementation.swift#L281))

File: Sources/Tracking/CustomerIOImplementation.swift [hunk 7]
```diff
@@ -135,4 +134,107 @@ class CustomerIOImplementation: CustomerIOInstance {
...
+     */
+    func getAndProcessTask(for task: QueueTaskMetadata) {
+        guard let taskDetail = backgroundQueue.getTaskDetail(task) else { return }
+        let taskData = taskDetail.data
+        var isProcessed = true
+        switch taskDetail.taskType {
+        case .trackDeliveryMetric:
+            // TODO: Segment doesn't provide this method by default needs to get added
+            // Remove isProcessed when the method is added
+            print("Track Delivery Metrics for in-app - Needs discussion")
```
Comment (Shahroz16) [inline Sources/Tracking/CustomerIOImplementation.swift:165]: its the same method as `track metric` for push, we just need to exclude `recipient` 

 ```
let properties: [String: Any] = metaData.mergeWith([
            "metric": event.rawValue,
            "deliveryId": deliveryId,
        ])
```

File: Sources/DataPipeline/DataPipeline.swift [hunk 8]
```diff
@@ -3,6 +3,12 @@ import Segment
 
 public protocol DataPipelineInstance: CustomerIOInstance {
     var analytics: Analytics { get }
+    func processIdentifyFromBGQ(identifier: String, body: [String: Any]?)
+    func processScreenEventFromBGQ(identifier: String, name: String, timestamp: String?, properties: [String: Any])
+    func processEventFromBGQ(identifier: String, name: String, timestamp: String?, properties: [String: Any])
+    func processDeleteTokenFromBGQ(identifier: String, token: String)
+    func processRegisterDeviceFromBGQ(identifier: String, token: String, attributes: [String: Any]?)
+    func processPushMetricsFromBGQ(token: String, event: Metric, deliveryId: String, timestamp: String, metaData: [String: Any])
```
Comment (Shahroz16) [inline Sources/DataPipeline/DataPipeline.swift:11]: we should have timestamps in here, commenting in the method i think we can get that value from.

File: Sources/Common/Background Queue/Queue.swift [hunk 9]
```diff
@@ -125,6 +141,31 @@ public class CioQueue: Queue {
         self.dateUtil = dateUtil
     }
 
+    // TODO: Write test case
+    public func getAllStoredTasks() -> [QueueTaskMetadata] {
+        storage.getInventory()
+    }
+
+    // TODO: Write test case
+    public func getTaskDetail(_ task: QueueTaskMetadata) -> (data: Data, taskType: QueueTaskType)? {
```
Comment (Shahroz16) [inline Sources/Common/Background Queue/Queue.swift:150]: we can probably get the timestamp from here?

```
        let createdAt = task.createdAt
```

File: Sources/DataPipeline/DataPipelineImplementation.swift (hunk 5 above)
Comment (Shahroz16) [reply Sources/DataPipeline/DataPipelineImplementation.swift:258]: These timestamps are from the metric model, we can have timestamp for events from `TaskMetaData`. I added the comment where i think we can probably get and return that information.

Comment (ami-aman) [reply Sources/DataPipeline/DataPipelineImplementation.swift:258]: That is `createdAt` which I believe might differ in terms of the actual `timestamp`. If that works for us then I can utilise and make updates in all the methods.

File: Apps/APN-UIKit/APN UIKit/View/LoginViewController.swift (hunk 6 above)
Comment (ami-aman) [reply Apps/APN-UIKit/APN UIKit/View/LoginViewController.swift:101]: This can be done in a separate PR as this is unrelated to this ticket.

File: Sources/Tracking/CustomerIOImplementation.swift (hunk 7 above)
Comment (ami-aman) [reply Sources/Tracking/CustomerIOImplementation.swift:170]: The calls are automatically going to `trackMetric`. What I have been thinking about is the use case of `trackDeliveryMetric` as I could not figure out a way to test this case. I added a "need discussion" comment because this is one of the possible cases in `QueueTaskType` but I had a hard time trying to reproduce this one.

File: Sources/DataPipeline/DataPipeline.swift (hunk 8 above)
Comment (ami-aman) [reply Sources/DataPipeline/DataPipeline.swift:11]: [Response](https://github.com/customerio/customerio-ios/pull/424#discussion_r1437854876)

File: Sources/Common/Background Queue/Queue.swift (hunk 9 above)
Comment (ami-aman) [reply Sources/Common/Background Queue/Queue.swift:148]: [Response](https://github.com/customerio/customerio-ios/pull/424#discussion_r1437854876)

File: Sources/Tracking/CustomerIO.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (ami-aman) [reply Sources/Tracking/CustomerIO.swift]: ✅

File: Sources/DataPipeline/DataPipelineImplementation.swift (hunk 5 above)
Comment (ami-aman) [reply Sources/DataPipeline/DataPipelineImplementation.swift]: Updated the code with `timestamp` as available, if not then `createdAt` from `QueueTaskMetadata` !!!

File: Sources/DataPipeline/DataPipelineImplementation.swift [hunk 10]
```diff
@@ -252,3 +251,88 @@ extension DataPipelineImplementation {
         return attributesPlugin
     }
 }
+
+// To process pending tasks in background queue
+// BGQ in each of the following methods refer to background queue
+extension DataPipelineImplementation {
+    func processIdentifyFromBGQ(identifier: String, timestamp: String, body: [String: Any]?) {
+        var identifyEvent = IdentifyEvent(userId: identifier, traits: nil)
+        identifyEvent.timestamp = timestamp
```
Comment (Shahroz16) [inline Sources/DataPipeline/DataPipelineImplementation.swift:260]: just make sure we have verified the format of timestamp that we are adding from BQ is same as one Analytics expect

File: Sources/DataPipeline/DataPipelineImplementation.swift [hunk 11]
```diff
@@ -252,3 +251,88 @@ extension DataPipelineImplementation {
...
+        analytics.process(event: trackRegisterTokenEvent)
+    }
+
+    func processPushMetricsFromBGQ(token: String, event: Metric, deliveryId: String, timestamp: String, metaData: [String: Any] = [:]) {
+        let properties: [String: Any] = metaData.mergeWith([
+            "metric": event.rawValue,
+            "deliveryId": deliveryId,
+            "recipient": token
+        ])
+//        if let token = deviceToken {
```
Comment (Shahroz16) [inline Sources/DataPipeline/DataPipelineImplementation.swift:329]: why do we have this commented-out version here?

File: Sources/Tracking/CustomerIOImplementation.swift [hunk 12]
```diff
@@ -135,4 +134,108 @@ class CustomerIOImplementation: CustomerIOInstance {
     ) {
         DataPipeline.shared.trackMetric(deliveryID: deliveryID, event: event, deviceToken: deviceToken)
     }
+
+    func handleQueueBacklog() {
+        let allStoredTasks = backgroundQueue.getAllStoredTasks()
+        if allStoredTasks.count <= 0 {
+            logger.info("No tasks pending in the background queue to be executed.")
```
Comment (Shahroz16) [inline Sources/Tracking/CustomerIOImplementation.swift:141]: Probably could use more detail, like its CIO migration queue etc so we dont confuse what queue is the log about

File: Tests/Tracking/CustomerIOImplementationTest.swift [hunk 13]
```diff
@@ -426,4 +428,56 @@ class CustomerIOImplementationTest: UnitTest {
...
+    func test_givenEmptyBacklog_expectNoTasksProcessed() {
+        backgroundQueueMock.getAllStoredTasksReturnValue = []
+        XCTAssertNotNil(implementation.handleQueueBacklog())
+        XCTAssertEqual(backgroundQueueMock.getAllStoredTasksCallsCount, 1)
+    }
+
+    func test_givenBacklog_expectTaskProcessed() {
+        var inventory: [QueueTaskMetadata] = []
+        let givenType = QueueTaskType.identifyProfile
+        let givenTask = IdentifyProfileQueueTaskData(identifier: String.random, attributesJsonString: "null")
```
Comment (Shahroz16) [inline Tests/Tracking/CustomerIOImplementationTest.swift:443]: we are only appending the same kind of task, it would have been great if we could have an additional test case where we verify each type of task and make sure its values are being received accordingly.

**Review by Shahroz16 — APPROVED** [review]
Looks good, some suggestions to increase confidence in migration tasks.

File: Tests/Tracking/CustomerIOImplementationTest.swift (hunk 13 above)
Comment (ami-aman) [reply Tests/Tracking/CustomerIOImplementationTest.swift:443]: I do not think that will change the functionality in any way., Since it is a test case and not actual implementation so I do not believe it will make any change!

File: Sources/Tracking/CustomerIOImplementation.swift (hunk 12 above)
Comment (ami-aman) [reply Sources/Tracking/CustomerIOImplementation.swift:141]: That makes sense.

File: Sources/DataPipeline/DataPipelineImplementation.swift (hunk 11 above)
Comment (ami-aman) [reply Sources/DataPipeline/DataPipelineImplementation.swift:329]: My bad ! I made this change to test a condition and didn't reverse it.

File: Sources/DataPipeline/DataPipelineImplementation.swift (hunk 10 above)
Comment (ami-aman) [reply Sources/DataPipeline/DataPipelineImplementation.swift:260]: Updated all timestamps to ISO format.

**Comment by ami-aman** [comment]
> Looks good, some suggestions to increase confidence in migration tasks.

I would still suggest to do a peer and group testing of this feature since this is the major one. I am confident in it's working as far as I have tested but I do agree that dev testing is not enough at times specially when the feature is an important one like this ! We can do a test run once the team is back next week of all the modules that have been done. It can be a kind of UAT !
